### PR TITLE
GNIP 94: Generic and pluggable OIDC SocialAccount Provider for GeoNode

### PR DESCRIPTION
**Pull request by afabiani**
References: https://github.com/GeoNode/geonode/issues/11185

## TODO
- [x] Extend the GNIP in order to recognize an optional attribute to set the `manager` role for a user at login time 
- [x] Documentation
- [x] Unit Tests

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

For all pull requests:

- [x] Confirm you have read the [contribution guidelines](https://github.com/GeoNode/geonode/blob/master/CONTRIBUTING.md) 
- [x] You have sent a Contribution Licence Agreement (CLA) as necessary (not required for small changes, e.g., fixing typos in the documentation)
- [x] Make sure the first PR targets the master branch, eventual backports will be managed later. This can be ignored if the PR is fixing an issue that only happens in a specific branch, but not in newer ones.

The following are required only for core and extension modules (they are welcomed, but not required, for contrib modules):
- [x] There is a ticket in https://github.com/GeoNode/geonode/issues describing the issue/improvement/feature (a notable exemption is, changes not visible to end-users)
- [x] The issue connected to the PR must have Labels and Milestone assigned
- [ ] PR for bug fixes and small new features are presented as a single commit
- [x] Commit message must be in the form "[Fixes #<issue_number>] Title of the Issue"
- [ ] New unit tests have been added covering the changes, unless there is an explanation on why the tests are not necessary/implemented
- [x] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [x] This PR passes the QA checks: black geonode && flake8 geonode
- [ ] Commits changing the **settings**, **UI**, **existing user workflows**, or adding **new functionality**, need to include documentation updates
- [ ] Commits adding **new texts** do use gettext and have updated .po / .mo files (without location infos)

**Submitting the PR does not require you to check all items, but by the time it gets merged, they should be either satisfied or inapplicable.**

### CODE REVIEW

**Review by giohappy — CHANGES_REQUESTED** [review]
@afabiani some tests failing

File: geonode/settings.py [hunk 1]
```diff
@@ -1956,56 +1955,64 @@ def get_geonode_catalogue_service():
...
+    "COMMON_FIELDS": {"email": "email", "last_name": "family_name", "first_name": "given_name"},
+    "ACCOUNT_CLASS": "allauth.socialaccount.providers.google.provider.GoogleAccount",
+    "ACCESS_TOKEN_URL": "https://oauth2.googleapis.com/token",
+    "AUTHORIZE_URL": "https://accounts.google.com/o/oauth2/v2/auth",
+    "ID_TOKEN_ISSUER": "https://accounts.google.com",
+    "OAUTH_PKCE_ENABLED": True,
+}
+
+SOCIALACCOUNT_PROVIDERS = {
+    "geonode_openid_connect": _AZURE_SOCIALACCOUNT_PROVIDER,
```
Comment (giohappy) [inline geonode/settings.py:2011]: @afabiani why aren't we including also the `_GOOGLE_SOCIALACCOUNT_PROVIDER` here, if Azure is configured by default?

File: geonode/utils.py [hunk 2]
```diff
@@ -17,10 +17,10 @@
 #
 #########################################################################
 
-import itertools
 import os
 import gc
 import re
+import imp
```
Comment (giohappy) [inline geonode/utils.py:23]: @afabiani the `imp` module is deprecated in favor of `importlib`, which is already used in several places in GeoNode.

File: geonode/settings.py (hunk 1 above)
Comment (afabiani) [reply geonode/settings.py]: @giohappy you need to choose one or another in this implementation. I rfactored this a bit so it is a bit more clear

**Review by giohappy — COMMENTED** [review]
@afabiani could we add some tests?
We might take inspiration from the allauth tests for the single providers, like [this](https://github.com/pennersr/django-allauth/blob/main/allauth/socialaccount/providers/google/tests.py)

**Comment by giohappy** [comment]
This would also fix https://github.com/GeoNode/geonode/issues/8814

**Comment by giohappy** [comment]
@afaI see that the extraction and promotion of the group manager have been implemented. However I don't see where a user is demoted in case at the next login we find that it's not a manager anymore.

**Comment by giohappy** [comment]
ok, I see that if a token has the claim for the groups, that user will be removed from the groups at each login, and reassigned to the groups reported inside the claims.

This could be problematic if the user is also assigned to a group manually since the membership will be deleted at the next login (if the same user has the groups declared in its token).
**We will have to add a note on this inside the documentation** @afabiani

**Review by giohappy — APPROVED** [review]

**Pull request merged**